### PR TITLE
feat: add validation layer for workflow

### PR DESCRIPTION
**Pull request by anujsinha3**
#79 

- added a validation.py file to validate command line arguments for non-negative numerical values
- added tests for validation.py

### CODE REVIEW

File: src/mawpy/utilities/validations.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+def check_non_negative(value):
+    try:
+        f_value = float(value)
+    except Exception as ex:
+        logger.error(ex)
+        raise argparse.ArgumentTypeError(f"{value} is not a valid non-negative numerical value")
+    if f_value < 0:
+        raise argparse.ArgumentTypeError(f"{value} is not a valid non-negative numerical value")
+    return f_value
```
Comment (lsetiawan) [inline src/mawpy/utilities/validations.py:15]: There are 2 type of check that need to happen here and probably 2 different exception. Using this `argparse` functionality to check for negative value is not preferred and I would consider this very hacky. You should defer the negative value check after the value is parsed by argparse and leave this for type checking only..

File: tests/test_validations.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+@given(st.floats(max_value=-1e-10, allow_infinity=False, allow_nan=False))
+def test_invalid_negative_value(value):
+    # Hypothesis generates various negative floats
+    try:
+        check_non_negative(value)
+    except ValueError:
+        pass  # Expected exception
+    else:
+        assert False, f"Expected Value Error for value: {value}"
```
Comment (lsetiawan) [inline tests/test_validations.py:24]: Please modify this to use pytest assertion for expected exception: https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions

File: tests/test_validations.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+from mawpy.utilities.validations import check_non_negative
+
+logger = logging.getLogger(__name__)
+
+
+@given(st.floats(min_value=0, allow_infinity=False, allow_nan=False))
+def test_valid_non_negative_values(value):
+    # Hypothesis generates various non-negative floats, including edge cases like 0.0
+    check_non_negative(value)
+    pass
```
Comment (lsetiawan) [inline tests/test_validations.py:13]: I see 2 tests for the `check_non_negative` please add test for `validate_input_args` as well.

File: src/mawpy/utilities/validations.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+from mawpy.constants import INPUT_ARGS_LIST
+
+
+def check_non_negative(value):
```
Comment (lsetiawan) [inline src/mawpy/utilities/validations.py:4]:
```suggestion
def check_non_negative(value: float):
```

File: src/mawpy/utilities/validations.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+    if value < 0:
+        raise ValueError(f"{value} is not a valid non-negative numerical value")
+
+
+def validate_input_args(**kwargs):
```
Comment (lsetiawan) [inline src/mawpy/utilities/validations.py:9]:
```suggestion
def validate_input_args(**kwargs: float) -> None:
```

Comment (anujsinha3) [reply src/mawpy/utilities/validations.py]: updated for now - but this might change if we add additional validation functions for different types within the parent function.

File: src/mawpy/utilities/validations.py (hunk 4 above)
Comment (anujsinha3) [reply src/mawpy/utilities/validations.py]: updated.

File: tests/test_validations.py (hunk 3 above)
Comment (anujsinha3) [reply tests/test_validations.py]: added

File: tests/test_validations.py (hunk 2 above)
Comment (anujsinha3) [reply tests/test_validations.py]: Thanks for this one, updated!

File: tests/test_validations.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+@given(st.floats(max_value=-1e-10))
+def test_validate_input_args_for_negative_arg_not_in_validation_list(abc):
+    # No Exception as Hypothesis generates negative floats, but the argument passed is not in the validation list.
+    validate_input_args(abc=abc)
+
+
+@given(st.floats(max_value=-1e-10, allow_infinity=False, allow_nan=False))
+def test_validate_input_args_negative_value(negative_float_value):
+    # Raises Value Error as the passed argument is in the validation list for negative value check.
+    pytest.raises(ValueError, validate_input_args, duration_constraint=negative_float_value)
```
Comment (lsetiawan) [inline tests/test_validations.py:39]:
```suggestion
    with pytest.raises(ValueError, match="is not a valid non-negative numerical value"):
        validate_input_args(duration_constraint=negative_float_value)
```

pytest raise check should be used as a context manager here

File: tests/test_validations.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+@given(st.floats(min_value=0, allow_infinity=False, allow_nan=False))
+def test_valid_non_negative_values(value):
+    # Hypothesis generates various non-negative floats, including edge cases like 0.0
+    check_non_negative(value)
+
+
+@given(st.floats(max_value=-1e-10, allow_infinity=False, allow_nan=False))
+def test_invalid_negative_value(value):
+    # Hypothesis generates various negative floats
+    pytest.raises(ValueError, check_non_negative, value)
```
Comment (lsetiawan) [inline tests/test_validations.py:20]:
```suggestion
    with pytest.raises(ValueError, match="is not a valid non-negative numerical value"):
        check_non_negative(value)
```

File: src/mawpy/utilities/validations.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
+from mawpy.constants import INPUT_ARGS_LIST
+
+
+def check_non_negative(value: float) -> None:
+    if value < 0:
+        raise ValueError(f"{value} is not a valid non-negative numerical value")
+
+
+def validate_input_args(**kwargs: float) -> None:
```
Comment (carlosgjs) [inline src/mawpy/utilities/validations.py:9]: Since this only validates non-negative values, I suggest renaming it to `validate_non_negative_args` and then just inlining the `<0`. This is simpler and clearer. 

But if we expect this will need to generalize further you could create a map of validation functions for each arg, eg:
```
validations = { 
   "duration": [check_non_negative, check_finite]
   "spatial_constraint: [partial(check_range, min=0.1, max=10.0)], 
}
```

**Review by carlosgjs — APPROVED** [review]
Please see inline suggestions

File: src/mawpy/utilities/validations.py (hunk 8 above)
Comment (anujsinha3) [reply src/mawpy/utilities/validations.py:15]: Updated as per suggestion.

File: src/mawpy/utilities/validations.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+    for k, val in kwargs.items():
+        if k in INPUT_ARGS_LIST:
+            check_non_negative(val)
```
[the remark this answers is not in the document]
Comment (anujsinha3) [reply src/mawpy/utilities/validations.py]: Thanks, Carlos! 
have added a check now that will ensure that some validation is performed when this function is called, if no validation is performed during the function call then an assertion error will be raised.

File: tests/test_validations.py (hunk 7 above)
Comment (anujsinha3) [reply tests/test_validations.py]: Thanks, Don! Updated as per suggestion
